Automattic/hostmgr
# Reduce SotoS3 dependency

The SotoS3 dependency is very complex and uses a lot of CPU to do its work.

This PR reduces our need for it by encapsulating it within the `libhostmgr` target. From there, it's only used to sign S3 requests, which are then processed using `URLSession` (wrapped by Alamofire). 

This makes download operations considerably faster and more efficient by avoiding all the overhead of SwiftNIO, instead relying on boring HTTP request handling. In the future, it'd be great to eliminate the rest of SotoS3 usage, but this gets us most of the way to where we want to be.

A big part of the work here is adopting async/await – this allows us to avoid polling and deep callback trees, making it much simpler to chain high-level operations together. A forthcoming PR will take advantage of this to provide higher-level methods that we can take advantage of on our Mac VM nodes.

Note: this PR builds on top of #39 and will be available for review once it lands.

**To Test**
Try running these commands, ensure they succeed:

- `swift run hostmgr sync --help`
- `swift run hostmgr sync authorized_keys` (note: this will create/overwrite the `authorized_keys` file on your local machine Run `rm -rf ~/.ssh/authorized_keys` to get rid of it after)
- `swift run hostmgr sync vm_images` (note: this will take a _really_ long time to run, feel free to ctrl+c to cancel it after ~10 seconds or so – the same code that handles the authorized keys sync is used under the hood so you don't have to wait for it to finish)
- `swift run hostmgr vm image remote list` (note: this should show a full list of VM images, regardless of the contents of the manifest)
- `swift run hostmgr vm image remote download xcode-14` (note: don't feel the need to wait for this to finish. As well, the progress bar is messed up – we'll tear it out in a subsequent PR to keep this one from getting _too_ unfocused).
- `swift run hostmgr benchmark` (note: it should show a help screen with a list of subcommands)
- `swift run hostmgr benchmark network` (note: once it starts printing data, things are working and you can close it down)
- `swift run hostmgr benchmark disk` (note: once it starts printing data, things are working and you can close it down)
